CNDB-11092 main: Fix IndexInputLeakDetector

adelapena opened the pull request:
Fix IndexInputLeakDetector, which has been mostly no-op for a long time. It detects opened index inputs that haven't been closed. It seemed originally designed to also detect attempts at closing index inputs that had already been closed, but the current usage seems to do that kind of duplicated closing, which is probably ok if the operation is idempotent.

pkolaczk commented on src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java:
```diff
@@ -55,13 +55,29 @@ public class IndexFileUtils
                                                                                               .build();
 
-    public static final IndexFileUtils instance = new IndexFileUtils();
+    private static final IndexFileUtils instance = new IndexFileUtils(defaultWriterOption);
+    private static IndexFileUtils overrideInstance = null;
```
It's a bit nitpicking, but is there any particular reason we use a combination of two variables instead of just marking the first one non-final and updating it in place? Then we wouldn't need to change all the stuff from `.instance` to `.instance()` (which is fine, but I'm worried it increases the conflict area in any merges we do e.g. wit apache). Is there any downside of using a single field?

adelapena replied:
Not sure why it was done this way, this comes from [a previous set of changes in `main-5.0`](https://github.com/datastax/cassandra/commit/047d33c904e6082708000698b4342d236b427f71#diff-6a43d8be03e631756ea0fd2a6fd92b6dd106ed1450ab6cabf40dd86d551bb47d). Agree on the single variable. Regarding having getters and setters, the only reason that occurs to me is perhaps having less visibility in the setter? In any case, I have changed it to the simplest approach.

By the way, regarding Apache we have CASSANDRA-19961 to update this there too, albeit in that case the problem is that some inputs are not being closed.

pkolaczk commented on test/unit/org/apache/cassandra/index/sai/disk/io/TrackingIndexFileUtils.java:
```diff
@@ -19,78 +19,114 @@
 public class TrackingIndexFileUtils extends IndexFileUtils
 {
     private final Map<TrackingIndexInput, String> openInputs = Collections.synchronizedMap(new HashMap<>());
+    private final Set<TrackingIndexInput> closedInputs = new HashSet<>();
```
bug: That should be synchronized set, as inputs can be closed by any threads. The `synchronized` on `close` does not cut it because it does not synchronize on this common utils object, but it synchronizes on the individual `IndexInput`. So two different index inputs closing at the same time would attempt to insert into this set concurrently.

adelapena replied:
Right, good catch, fixed.

pkolaczk reviewed (changes requested):
Overall looks good, but I have one concern about synchronization of the set.

adelapena commented:
Mmh, it seems the fixed detector has found some leaks. Looking into it.

adelapena commented:
The found leak was due to `SortedTermsReader.Cursor` not closing `blockOffsets`, which is done in `main-5.0` but not here in `main`. Fixed.

adelapena commented:
It seems the issues reported by Sonar are due to the changes around `IndexFileUtils` discussed above.

adelapena commented:
It seems some minor merge conflicts have appeared since addressing the review feedback last week. I have pushed a rebased branch.

@pkolaczk are you ok with the changes?
